File: src/main/java/com/assets/board/repository/ExchangeRateRepository.java
package com.assets.board.repository;

import com.assets.board.entity.ExchangeRate;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

public interface ExchangeRateRepository extends JpaRepository<ExchangeRate, Long> {

    Optional<ExchangeRate> findByDate(LocalDate date);

    List<ExchangeRate> findByDateBetween(LocalDate start, LocalDate end);

    @Query("SELECT e FROM ExchangeRate e WHERE e.date = " +
            "(SELECT MAX(e2.date) FROM ExchangeRate e2 WHERE e2.date <= :date)")
    Optional<ExchangeRate> findLatestBeforeOrOn(@Param("date") LocalDate date);
}

File: src/main/java/com/assets/board/repository/ExchangeRateRepository.kt
package com.assets.board.repository

import com.assets.board.entity.ExchangeRate
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.time.LocalDate
import java.util.*

@Repository
interface ExchangeRateRepository : JpaRepository<ExchangeRate?, Long?> {
    fun findByDate(date: LocalDate?): ExchangeRate?

    fun findByDateBetween(start: LocalDate?, end: LocalDate?): MutableList<ExchangeRate?>?

    @Query(
        "SELECT e FROM ExchangeRate e WHERE e.date = " +
                "(SELECT MAX(e2.date) FROM ExchangeRate e2 WHERE e2.date <= :date)"
    )
    fun findLatestBeforeOrOn(@Param("date") date: LocalDate?): Optional<ExchangeRate?>?
}
